packages/aragon-cli/src/commands/apm_cmds/util/generate-artifact.js

```diff
@@ -0,0 +1,215 @@
+        web3.eth.abi.encodeFunctionSignature(f.sig)
+    )
+  })
+}
+
+async function deprecatedFunctions(apm, artifact, web3, reporter) {
+  let deprecated = {}
+  try {
+    const deprecatedFunctionsSig = new Set()
+    const versions = await apm.getAllVersions(artifact.appName)
+    let lastMajor = -1
+    versions.reverse().forEach(async version => {
+      let deprecatedOnVersion = []
+      // iterate on major versions
+      if (getMajor(version.version) !== lastMajor) {
+        lastMajor = getMajor(version.version)
+        if (version.functions) {
+          version.functions.forEach(f => {
+            if (
+              !artifact.functions.some(obj => obj.sig === f.sig) &&
+              !deprecatedFunctionsSig.has(f.sig)
+            ) {
+              deprecatedOnVersion.push(f)
+              deprecatedFunctionsSig.add(f.sig)
+            }
+          })
+          if (deprecatedOnVersion.length) {
+            deprecated[`${lastMajor}.0.0`] = deprecatedOnVersion
+            decorateFunctionsWithAbi(deprecatedOnVersion, version.abi, web3)
+            deprecatedOnVersion = []
+          }
+        } else {
+          reporter.warning(
+            `Cannot find artifacts for version ${version.version}  in aragonPM repo. Please make sure the package was published and IPFS or your HTTP server running.`
+          )
+          return deprecated
+        }
+      }
+    })
+    return deprecated
+  } catch (e) {
+    // Catch ENS error on first version
+  }
+}
+
+async function generateApplicationArtifact(
+  cwd,
+  apm,
+  outputPath,
+  module,
+  deployArtifacts,
+  web3,
+  reporter
+) {
+  let artifact = Object.assign({}, module)
+  const contractPath = artifact.path
+  const contractInterfacePath = path.resolve(
+    cwd,
+    'build/contracts',
+    path.basename(contractPath, '.sol') + '.json'
+  )
+
+  // Set `appId`
+  artifact.appId = namehash.hash(artifact.appName)
+
+  // Set ABI
+  const contractInterface = await readJson(contractInterfacePath)
+  artifact.abi = contractInterface.abi
+
+  if (deployArtifacts) {
+    artifact.deployment = deployArtifacts
+    if (deployArtifacts.flattenedCode) {
+      fs.writeFileSync(
+        path.resolve(outputPath, SOLIDITY_FILE),
+        artifact.deployment.flattenedCode
+      )
+      artifact.deployment.flattenedCode = `./${SOLIDITY_FILE}`
+    }
+  }
+
+  // Analyse contract functions and returns an array
+  // > [{ sig: 'transfer(address)', role: 'X_ROLE', notice: 'Transfers..'}]
+  artifact.functions = await extract(path.resolve(cwd, artifact.path))
+  // extract abi for each function
+  // > [{ sig: , role: , notice: , abi: }]
+  decorateFunctionsWithAbi(artifact.functions, artifact.abi, web3)
+
+  // Consult old (major) version's artifacts and return an array
+  // of deprecated functions per version
+  // > "deprecated": { "1.0.0": [{}], "2.0.0": [{}] }
+  artifact.deprecated = await deprecatedFunctions(apm, artifact, web3, reporter)
```

Comment:
@0xGabi What do you think about making this `deprecatedFunctions` instead? `deprecated` sounds a bit vague (and might be taken to meant that the entire app is deprecated).